### PR TITLE
Ports Drastic Lag Mitigation Subsystem: SSlag_switch

### PR DESCRIPTION
**Pull request by Chirynnn**
## About The Pull Request
Ports https://github.com/tgstation/tgstation/pull/59717/, adding a new admin tool for hopefully mitigating lag to an extent during busy shifts. Unfortunately, I couldn't find a method to automatically add TRAIT_BYPASS_MEASURES to admin observers/players, so they are not exempt from the bottom options as of yet, I hope someone can help me with this.

## Why It's Good For The Game
Admeme tool for less lag good

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

https://user-images.githubusercontent.com/77589317/224504805-870c7db9-4bb9-4715-926b-20c173f8f369.mp4

</details>

## Changelog
:cl:

admin: Added a lag mitigation panel in the Admin tab
/:cl:

### CODE REVIEW

**Comment by itsmeow** [comment]
you can iterate over the player list, check if they have a `holder`, and then add the trait

**Comment by itsmeow** [comment]
or just attached it to holder init and destroy

**Comment by Chirynnn** [comment]
i have no idea what that means but you got it

File: code/__HELPERS/icons.dm [hunk 1]
```diff
@@ -1276,6 +1276,8 @@ GLOBAL_LIST_INIT(freon_color_matrix, list("#2E5E69", "#60A2A8", "#A1AFB1", rgb(0
 /proc/icon2html(atom/thing, client/target, icon_state, dir = SOUTH, frame = 1, moving = FALSE, sourceonly = FALSE)
 	if (!thing)
 		return
+	if(SSlag_switch.measures[DISABLE_USR_ICON2HTML] && usr && !HAS_TRAIT(usr, TRAIT_BYPASS_MEASURES))
```
Comment (PowerfulBacon) [inline code/__HELPERS/icons.dm:1279]: Use target.mob instead of usr (Make sure to null check)

File: code/__HELPERS/icons.dm [hunk 2]
```diff
@@ -1381,6 +1383,8 @@ GLOBAL_LIST_INIT(freon_color_matrix, list("#2E5E69", "#60A2A8", "#A1AFB1", rgb(0
 /proc/costly_icon2html(thing, target, sourceonly = FALSE)
 	if (!thing)
 		return
+	if(SSlag_switch.measures[DISABLE_USR_ICON2HTML] && usr && !HAS_TRAIT(usr, TRAIT_BYPASS_MEASURES))
```
Comment (PowerfulBacon) [inline code/__HELPERS/icons.dm:1386]: Same here, use target instead of usr

File: code/controllers/subsystem/lag_switch.dm [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	var/static/list/measures[MEASURES_AMOUNT]
+	/// List of measures that toggle automatically
+	var/list/auto_measures = list(DISABLE_GHOST_ZOOM_TRAY, DISABLE_RUNECHAT, DISABLE_USR_ICON2HTML)
+	/// Timer ID for the automatic veto period
+	var/veto_timer_id
+	/// Cooldown between say verb uses when slowmode is enabled
+	var/slowmode_cooldown = 3 SECONDS
+
+/datum/controller/subsystem/lag_switch/Initialize(start_timeofday)
+	for(var/i = 1, i <= measures.len, i++)
```
Comment (PowerfulBacon) [inline code/controllers/subsystem/lag_switch.dm:20]:
```suggestion
	for(var/i = 1 to length(measures))
```

File: code/controllers/subsystem/lag_switch.dm [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+	/// Amount of connected clients at which the Lag Switch should engage, set via config or admin panel
+	var/trigger_pop = INFINITY - 1337
+	/// List of bools corresponding to code/__DEFINES/lag_switch.dm
+	var/static/list/measures[MEASURES_AMOUNT]
+	/// List of measures that toggle automatically
+	var/list/auto_measures = list(DISABLE_GHOST_ZOOM_TRAY, DISABLE_RUNECHAT, DISABLE_USR_ICON2HTML)
+	/// Timer ID for the automatic veto period
+	var/veto_timer_id
+	/// Cooldown between say verb uses when slowmode is enabled
+	var/slowmode_cooldown = 3 SECONDS
```
Comment (PowerfulBacon) [inline code/controllers/subsystem/lag_switch.dm:17]: Slowmode should be like 0.5 seconds imo, 3 seconds affects gameplay way too much

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]

**Pull request closed without merging**